### ISSUE TITLE
Add yaml tags to schema in kairos-sdk

### ISSUE THREAD

**Issue by nianyush** [kairos-io/kairos#2733]
Right now, it only has json tags which is not usable to unmarshal/marshal kairos cloud configs
https://github.com/kairos-io/kairos-sdk/blob/main/schema/install_schema.go#L10

**Comment by mauromorales** [comment]
@kairos-io/maintainers this is related to #2539 and the fact that we have 2 mechanisms one for the actual reading and installing of the config and one just for validating (the schema). I'd rather see us work on finding a single source of truth than growing more features on the schema. WDYT?

**Comment by Itxaka** [comment]
> @kairos-io/maintainers this is related to #2539 and the fact that we have 2 mechanisms one for the actual reading and installing of the config and one just for validating (the schema). I'd rather see us work on finding a single source of truth than growing more features on the schema. WDYT?

I think that while that its good, just adding the tags will probably unblock issues for @nianyush so we can do this first as its quick and easy and then work in a better merged solution. We can ship this as part of 3.1.1 and then work with no hurries as there is a patch for it. WDYT?

**Comment by mauromorales** [comment]
I would not block this from being done, but this refactoring has been parked for more than a year and IMO if we don't have a reason to do it, it will just keep waiting for another year. So, we can apply the fix, but then we continue having the root problem, and tomorrow another enhancement will come and in the end we will be putting more work into workarounds than the actual solution. We also already had a case where installation structs and schema structs differed, and we didn't notice, so I'm just trying to raise a flag here to say that this should not stay in the background if possible, but again, we can, of course, make the change if that helps unblock.

**Comment by Itxaka** [comment]
@nianyush is this high prio? like would it unblock something on your side? or there is a workaround in place that solves the issue and doesnt have any side effects?

**Comment by santhoshdaivajna** [comment]
yes, this one is important. We are adding support for _user-data validation_ , this will help validating the install schema along with stylus attributes.

**Comment by mauromorales** [comment]
@Itxaka @santhoshdaivajna one thing to keep in mind is that while this will allow generating yaml, it doesn't mean it will be 100% valid, since the source of truth atm is the Install struct. We fixed the last known issue that made them different, but I cannot guarantee that they haven't diverged meanwhile. So it might be worth doing that validation while we work on this ticket.

**Comment by mauromorales** [comment]
@nianyush I just looked at the suggested PR, but I see there are a few cases that would have the exact problem I was describing before and would not be solved by simply adding yaml tags. Please give an in-depth explanation of what this issue is you are trying to solve. Just requesting the yaml tags is to vague for us to evaluate what is the correct path forwards.

**Comment by nianyush** [comment]
@mauromorales in our code, we will extend kairos install struct with our own fields to create a definition for our cloudconfigs. And right now, we are trying to do validation similar to what kairos does [here](https://github.com/kairos-io/kairos-sdk/blob/main/schema/validate.go). But the problem is that if we want to also support kairos fields validation, we cannot use the install struct inside kairos-agent anymore since it does not have those json fields. And it seems like if we can switch to the install_schema under kairos-sdk and adding yaml tags there, it will solve both problems: normal cloudconfig yaml marshaling and support for validation of kairos fields(if we just use the same json schema library as kairos-sdk)

**Comment by mauromorales** [comment]
@nianyush the problem I see with this, correct me if I'm wrong is that the install struct and the install schema do not map one to one 100%, for example the grub options on the install struct are a map[string]string, while on the install schema they are actual attributes of the struct. This is because on the validation, we care about those specific attributes and don't care to validate anything else but if you try to get back to yaml from a schema, you will have information that is lost.

Just to be on the same page, when I talk about the install struct I talk about https://github.com/kairos-io/kairos-agent/blob/main/pkg/config/config.go#L36. In the end, this is the only one that can marshal and unmarshall cloud configs.

**Comment by nianyush** [comment]
Yes @mauromorales i fully understand this. And i know you guys are talking about to move those two into one place. But to unblock us in this release, can we just validate install schema for once and use it? Tthen we can move to the new struct you guys proposed in the future since for us we wont update kairos version quite often, its often strictly tied with our release so once we validate the mapping of a specific version of kairos-sdk, then it should be good for us

**Comment by mauromorales** [comment]
@nianyush for me the problem is this part of the request 

> normal cloudconfig yaml marshaling and support for validation of kairos fields(if we just use the same json schema library as kairos-sdk

if we put the yaml tags there, it won't produce correct cloud-configs, and having it on the sdk means that users can consume it with the wrong idea of what it does. There's a reason why InstallStruct exists (to marshall and unmarshall cloud-configs) and one why the InstallSchema exists (just to check on certain values that we care about atm, like if a user was created, but it's not the actual source of truth, this is way this one specificaly only does json)

For you, would it be an option to consume the yaml the way we do with https://github.com/kairos-io/kairos-sdk/blob/main/schema/validate.go#L52? 

Or alternatively, as you also mention, you know that this will change, and you will have to adapt, why not introduce those structs on your side? That way we don't block you, but we also don't give the wrong intention with the code here until we decide how to address it properly

**Comment by nianyush** [comment]
@mauromorales yeah i think your point is correct. We will copy kairos install structs to our code for now and wait for kairos to merge those structs into one place and consume it later.

**Comment by mauromorales** [comment]
Closing this as per our conversation, but you are welcome to re-open if in case something still needs to be discussed.
